README.md
## Laravel DarkSky
[![Latest Stable Version](https://poser.pugx.org/naughtonium/laravel-dark-sky/v/stable)](https://packagist.org/packages/naughtonium/laravel-dark-sky)
[![Software License][ico-license]](LICENSE.md)
[![Total Downloads](https://poser.pugx.org/naughtonium/laravel-dark-sky/downloads)](https://packagist.org/packages/naughtonium/laravel-dark-sky)

This provides a Laravel style wrapper for the DarkSky api. For more information regarding request and response formats, visit: https://darksky.net/dev/docs


### Install

Require this package with composer using the following command:

``` bash
$ composer require naughtonium/laravel-dark-sky
```


After updating composer, add the service provider to the `providers` array in `config/app.php`

```php
Naughtonium\LaravelDarkSky\LaravelDarkSkyServiceProvider::class,
```

To register a facade accessor, add the following to `config/app.php` `aliases` array
```php
'DarkSky' => \Naughtonium\LaravelDarkSky\Facades\DarkSky::class,
```

### Configuration

Add the following line to the .env file:

```sh
DARKSKY_API_KEY=<your_darksky_api_key>
```


### Usage
For full details of response formats, visit: https://darksky.net/dev/docs/response

#### Required
##### location(lat, lon)
Pass in latitude and longitude coordinates for a basic response
``` php
DarkSky::location(lat, lon)->get();
```
#### Optional Parameters
For full details of optional parameters, visit: https://darksky.net/dev/docs/forecast

##### excludes([]) / includes([])
Specify which data blocks to exclude/include to reduce data transfer
```php
DarkSky::location(lat, lon)->excludes(['minutely','hourly', 'daily', 'alerts', 'flags'])->get();
DarkSky::location(lat, lon)->includes(['currently'])->get();
// Same output
```

##### atTime(t)
Pass in a unix timestamp to get forecast for that time.
Note: the timezone is relative to the given location

``` php
DarkSky::location(lat, lon)->atTime(timestamp)->get();
```
##### language(l)
Specify a language for text based responses
``` php
DarkSky::location(lat, lon)->language(lang)->get();
```
##### units(u)
Specify units for unit based responses
``` php
DarkSky::location(lat, lon)->units(units)->get();
```
##### extend()
Extend the "hourly" response from 48 to 168 hours.
Note: Does not work if used with an atTime() timestamp.
Please see: https://darksky.net/dev/docs/time-machine
``` php
DarkSky::location(lat, lon)->extend()->get();
```

#### Helpers
The following are shorthand helpers to add readability equal to using includes() with only one parameter. Note: only one may be used per query and only temperature specific data is returned
```php
->currently()
->minutely()
->hourly()
->daily()
->flags()
```
For example, these two statements are the same
```php
DarkSky::location(lat, lon)->hourly()
DarkSky::location(lat, lon)->includes(['hourly'])->get()->hourly
```
### Credits

- [Jack Naughton][link-author]
- [All Contributors][link-contributors]

### License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[link-author]: https://github.com/holiehandgrenade
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `laravel-dark-sky` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


prefill.php
<?php
define('COL_DESCRIPTION', 0);
define('COL_HELP', 1);
define('COL_DEFAULT', 2);

$fields = [
    'author_name' =>            ['Your name',             '',                                                ''],
    'author_github_username' => ['Your Github username',  '<username> in https://github.com/naughtonium',       ''],
    'author_email' =>           ['Your email address',    '',                                                ''],
    'author_twitter' =>         ['Your twitter username', '',                                                '@{author_github_username}'],
    'author_website' =>         ['Your website',          '',                                                'http://naughton.io'],

    'package_vendor' =>         ['Package vendor',        '<vendor> in https://github.com/vendor/package',   '{naughtonium}'],
    'package_name' =>           ['Package name',          '<package> in https://github.com/vendor/package',  ''],
    'package_description' =>    ['Package very short description',   '',                                     ''],

    'psr4_namespace' =>         ['PSR-4 namespace',       'usually, Vendor\\Package',                        '{naughtonium}\\{laravel-dark-sky}'],
];

$values = [
    'psr4_namespace' => 'naughtonium/laravel-dark-sky'
];

$replacements = [
    ':vendor\\\\:package_name\\\\' => function () use(&$values) { return str_replace('\\', '\\\\', $values['psr4_namespace']) . '\\\\'; },
    ':author_name'                 => function () use(&$values) { return $values['author_name']; },
    ':author_username'             => function () use(&$values) { return $values['author_github_username']; },
    ':author_website'              => function () use(&$values) { return $values['author_website'] ?: ('https://github.com/' . $values['author_github_username']); },
    ':author_email'                => function () use(&$values) { return $values['author_email'] ?: ($values['author_github_username'] . '@example.com'); },
    ':vendor'                      => function () use(&$values) { return $values['package_vendor']; },
    ':package_name'                => function () use(&$values) { return $values['package_name']; },
    ':package_description'         => function () use(&$values) { return $values['package_description']; },
    'League\\Skeleton'             => function () use(&$values) { return $values['psr4_namespace']; },
];

function read_from_console ($prompt) {
    if ( function_exists('readline') ) {
        $line = trim(readline($prompt));
        if (!empty($line)) {
            readline_add_history($line);
        }
    } else {
        echo $prompt;
        $line = trim(fgets(STDIN));
    }
    return $line;
}

function interpolate($text, $values)
{
    if (!preg_match_all('/\{(\w+)\}/', $text, $m)) {
        return $text;
    }
    foreach ($m[0] as $k => $str) {
        $f = $m[1][$k];
        $text = str_replace($str, $values[$f], $text);
    }
    return $text;
}

$modify = 'n';
do {
    if ($modify == 'q') {
        exit;
    }

    $values = [];

    echo "----------------------------------------------------------------------\n";
    echo "Please, provide the following information:\n";
    echo "----------------------------------------------------------------------\n";
    foreach ($fields as $f => $field) {
        $default = isset($field[COL_DEFAULT]) ? interpolate($field[COL_DEFAULT], $values): '';
        $prompt = sprintf(
            '%s%s%s: ',
            $field[COL_DESCRIPTION],
            $field[COL_HELP] ? ' (' . $field[COL_HELP] . ')': '',
            $field[COL_DEFAULT] !== '' ? ' [' . $default . ']': ''
        );
        $values[$f] = read_from_console($prompt);
        if (empty($values[$f])) {
            $values[$f] = $default;
        }
    }
    echo "\n";

    echo "----------------------------------------------------------------------\n";
    echo "Please, check that everything is correct:\n";
    echo "----------------------------------------------------------------------\n";
    foreach ($fields as $f => $field) {
        echo $field[COL_DESCRIPTION] . ": $values[$f]\n";
    }
    echo "\n";
} while (($modify = strtolower(read_from_console('Modify files with these values? [y/N/q] '))) != 'y');
echo "\n";

$files = array_merge(
    glob(__DIR__ . '/*.md'),
    glob(__DIR__ . '/*.xml.dist'),
    glob(__DIR__ . '/composer.json'),
    glob(__DIR__ . '/src/*.php'),
    glob(__DIR__ . '/tests/*.php')
);
foreach ($files as $f) {
    $contents = file_get_contents($f);
    foreach ($replacements as $str => $func) {
        $contents = str_replace($str, $func(), $contents);
    }
    file_put_contents($f, $contents);
}

echo "Done.\n";
echo "Now you should remove the file '" . basename(__FILE__) . "'.\n";


src/DarkSky.php
<?php
/**
 * Created by PhpStorm.
 * User: jack
 * Date: 6/13/17
 * Time: 3:26 PM
 */

namespace Naughtonium\LaravelDarkSky;


class DarkSky
{
    protected $apiKey;
    protected $endpoint = 'https://api.darksky.net/forecast/';
    protected $lat;
    protected $lon;
    protected $timestamp = null;
    protected $params = [];
    protected $excludeables = ['currently', 'minutely', 'hourly', 'daily', 'alerts', 'flags'];

    /**
     * DarkSky constructor.
     */
    public function __construct()
    {
        $this->apiKey = config('darksky.apikey');
        $this->endpoint = $this->endpoint . $this->apiKey;
    }

    /**
     * Sets the latitude and longitude. Must be set
     *
     * @param $lat
     * @param $lon
     * @return $this
     */
    public function location($lat, $lon)
    {
        $this->lat = $lat;
        $this->lon = $lon;
        return $this;
    }

    /**
     * Builds the endpoint url and sends the get request
     *
     * @return mixed
     */
    public function get()
    {
        $url = $this->endpoint  . '/' . $this->lat . ',' . $this->lon;

        if ($this->timestamp) {
            $url .= ',' . $this->timestamp;
        }

        $client = new \GuzzleHttp\Client();

        return json_decode($client->get($url, [
            'query' => $this->params,
        ])->getBody());
    }

    /**
     * Sets the exclude query parameter by taking array of exclusions
     *
     * @param $blocks
     * @return $this
     */
    public function excludes($blocks)
    {
        $this->params['exclude'] = implode(',', $blocks);
        return $this;
    }

    /**
     * Sets the exclude query parameter by taking an arry of inclusions
     *
     * @param $blocks
     * @return $this
     */
    public function includes($blocks)
    {
        $this->params['exclude'] = implode(',', array_diff($this->excludeables, $blocks));
        return $this;
    }

    /**
     * Specifies a unix timestamp for non-current forecasts
     * See: https://darksky.net/dev/docs/time-machine
     *
     * @param $timestamp
     * @return $this
     */
    public function atTime($timestamp)
    {
        $this->timestamp = $timestamp;
        return $this;
    }

    /**
     * Extends the number of hours returned in hourly from 48 to 168
     *
     * @return $this
     */
    public function extend()
    {
        $this->params['extend'] = 'hourly';
        return $this;
    }

    /**
     * Sets the return language
     *
     * @param $lang
     * @return $this
     */
    public function language($lang)
    {
        $this->params['lang'] = $lang;
        return $this;
    }

    /**
     * Sets the return units
     *
     * @param $unit
     * @return $this
     */
    public function units($unit)
    {
        $this->params['units'] = $unit;
        return $this;
    }

    ///////////////////////////////////////////////////////////////////
    //////////////////////////// HELPERS //////////////////////////////
    ///////////////////////////////////////////////////////////////////

    /**
     * Appends include for currently
     * 
     * @return $this
     */
    public function currently()
    {
        return $this->includes(['currently'])->get()->currently;
    }

    /**
     * Appends include for minutely
     *
     * @return $this
     */
    public function minutely()
    {
        return $this->includes(['minutely'])->get()->minutely;
    }

    /**
     * Appends include for hourly
     *
     * @return $this
     */
    public function hourly()
    {
        return $this->includes(['hourly'])->get()->hourly;
    }

    /**
     * Appends include for daily
     *
     * @return $this
     */
    public function daily()
    {
        return $this->includes(['daily'])->get()->daily;
    }

    /**
     * Appends include for alerts
     *
     * @return $this
     */
    public function alerts()
    {
        return $this->includes(['alerts'])->get()->alerts;
    }

    /**
     * Appends include for flags
     *
     * @return $this
     */
    public function flags()
    {
        return $this->includes(['flags'])->get()->flags;
    }
}